File: mechanical/src/main/scala/resonantinduction/mechanical/energy/grid/PartMechanical.java
package resonantinduction.mechanical.energy.grid;

import java.util.ArrayList;
import java.util.List;

import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.Item;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.MovingObjectPosition;
import net.minecraftforge.common.ForgeDirection;
import resonant.api.grid.INode;
import resonant.api.grid.INodeProvider;
import resonant.core.ResonantEngine;
import codechicken.lib.data.MCDataInput;
import codechicken.lib.data.MCDataOutput;
import codechicken.multipart.ControlKeyModifer;
import codechicken.multipart.JCuboidPart;
import codechicken.multipart.JNormalOcclusion;
import codechicken.multipart.TFacePart;

/** We assume all the force acting on the gear is 90 degrees.
 * 
 * @author Calclavia */
public abstract class PartMechanical extends JCuboidPart implements JNormalOcclusion, TFacePart, INodeProvider
{
    /** Node that handles mechanical action of the machine */
    public MechanicalNode node;
    protected double prevAngularVelocity;    
    int ticks = 0;
    /** Packets */
    boolean markPacketUpdate = false;
    /** Simple debug external GUI */
    MechanicalNodeFrame frame = null;

    /** Side of the block this is placed on */
    public ForgeDirection placementSide = ForgeDirection.UNKNOWN;

    public int tier;

    public void preparePlacement(int side, int itemDamage)
    {
        this.placementSide = ForgeDirection.getOrientation((byte) (side));
        this.tier = itemDamage;
    }

    @Override
    public void update()
    {
        ticks++;
        if (ticks >= Long.MAX_VALUE)
        {
            ticks = 0;
        }
        
        //Make sure to update on both sides
        this.node.update();
        
        if (!world().isRemote)
        {
            checkClientUpdate();          
        }
        if (frame != null)
        {
            frame.update();
        }
        super.update();
    }

    @Override
    public boolean activate(EntityPlayer player, MovingObjectPosition hit, ItemStack itemStack)
    {
        if (ResonantEngine.runningAsDev)
        {
            if (itemStack != null && !world().isRemote)
            {
                if (itemStack.getItem().itemID == Item.stick.itemID)
                {
                    //Set the nodes debug mode
                    if (ControlKeyModifer.isControlDown(player))
                    {
                        //Opens a debug GUI
                        if (frame == null)
                        {
                            frame = new MechanicalNodeFrame(this);
                            frame.showDebugFrame();
                        } //Closes the debug GUI
                        else
                        {
                            frame.closeDebugFrame();
                            frame = null;
                        }
                    }
                }
            }
        }
        return super.activate(player, hit, itemStack);
    }

    public void checkClientUpdate()
    {
        if (Math.abs(prevAngularVelocity - node.angularVelocity) >= 0.1)
        {
            prevAngularVelocity = node.angularVelocity;
            sendRotationPacket();
        }
    }

    @Override
    public INode getNode(Class<? extends INode> nodeType, ForgeDirection from)
    {
        if (nodeType.isAssignableFrom(node.getClass()))
            return node;
        return null;
    }

    @Override
    public void onWorldJoin()
    {
        node.reconstruct();
    }

    @Override
    public void onWorldSeparate()
    {
        node.deconstruct();
        if (frame != null)
        {
            frame.closeDebugFrame();
        }
    }

    /** Packet Code. */
    public void sendRotationPacket()
    {
        if (world() != null && !world().isRemote)
        {
            getWriteStream().writeByte(1).writeDouble(node.angularVelocity);
        }
    }

    /** Packet Code. */
    @Override
    public void read(MCDataInput packet)
    {
        read(packet, packet.readUByte());
    }

    public void read(MCDataInput packet, int packetID)
    {
        switch (packetID)
        {
            case 0:
                load(packet.readNBTTagCompound());
                break;
            case 1:
                node.angularVelocity = packet.readDouble();
                break;
        }
    }

    @Override
    public void readDesc(MCDataInput packet)
    {
        packet.readByte();
        load(packet.readNBTTagCompound());
    }

    @Override
    public void writeDesc(MCDataOutput packet)
    {
        packet.writeByte(0);
        NBTTagCompound nbt = new NBTTagCompound();
        save(nbt);
        packet.writeNBTTagCompound(nbt);
    }

    @Override
    public int redstoneConductionMap()
    {
        return 0;
    }

    @Override
    public boolean solid(int arg0)
    {
        return true;
    }

    @Override
    public void load(NBTTagCompound nbt)
    {
        placementSide = ForgeDirection.getOrientation(nbt.getByte("side"));
        tier = nbt.getByte("tier");
        node.load(nbt);
    }

    @Override
    public void save(NBTTagCompound nbt)
    {
        nbt.setByte("side", (byte) placementSide.ordinal());
        nbt.setByte("tier", (byte) tier);
        node.save(nbt);
    }

    protected abstract ItemStack getItem();

    @Override
    public Iterable<ItemStack> getDrops()
    {
        List<ItemStack> drops = new ArrayList<ItemStack>();
        drops.add(getItem());
        return drops;
    }

    @Override
    public ItemStack pickItem(MovingObjectPosition hit)
    {
        return getItem();
    }

    public universalelectricity.api.vector.Vector3 getPosition()
    {
        return new universalelectricity.api.vector.Vector3(x(), y(), z());
    }

    @Override
    public String toString()
    {
        return "[PartMech]" + x() + "x " + y() + "y " + z() + "z " + getSlotMask() + "s ";
    }
}

File: src/main/scala/resonantinduction/mechanical/mech/PartMechanical.scala
package resonantinduction.mechanical.mech

import codechicken.lib.data.{MCDataInput, MCDataOutput}
import codechicken.multipart._
import net.minecraft.nbt.NBTTagCompound
import net.minecraftforge.common.util.ForgeDirection
import resonant.lib.grid.UpdateTicker
import resonant.lib.transform.vector.VectorWorld
import resonantinduction.core.prefab.part.connector.{PartAbstract, TPartNodeProvider}
import resonantinduction.mechanical.mech.grid.NodeMechanical

/** We assume all the force acting on the gear is 90 degrees.
  *
  * @author Calclavia */
abstract class PartMechanical extends PartAbstract with JNormalOcclusion with TFacePart with TPartNodeProvider with TCuboidPart
{
  /** Node that handles resonantinduction.mechanical action of the machine */
  private var _mechanicalNode: NodeMechanical = null

  def mechanicalNode = _mechanicalNode

  def mechanicalNode_=(mech: NodeMechanical)
  {
    _mechanicalNode = mech
    mechanicalNode.onVelocityChanged = () => UpdateTicker.world.enqueue(() => sendPacket(1))
    nodes.add(mechanicalNode)
  }

  /** Side of the block this is placed on */
  var placementSide: ForgeDirection = ForgeDirection.UNKNOWN

  /** The tier of this mechanical part */
  var tier = 0

  def preparePlacement(side: Int, itemDamage: Int)
  {
    this.placementSide = ForgeDirection.getOrientation((side).asInstanceOf[Byte])
    this.tier = itemDamage
  }

  override def write(packet: MCDataOutput, id: Int)
  {
    super.write(packet, id)

    id match
    {
      case 0 =>
        val tag = new NBTTagCompound
        save(tag)
        packet.writeNBTTagCompound(tag)
      case 1 => packet.writeFloat(mechanicalNode.angularVelocity.toFloat)
      case 2 => packet.writeFloat(mechanicalNode.prevAngle.toFloat)
    }
  }

  override def read(packet: MCDataInput, id: Int)
  {
    super.read(packet, id)

    id match
    {
      case 0 =>
        load(packet.readNBTTagCompound())
      case 1 => mechanicalNode.angularVelocity = packet.readFloat()
      case 2 => mechanicalNode.prevAngle = packet.readFloat()
    }
  }

  override def redstoneConductionMap: Int = 0

  override def solid(arg0: Int): Boolean = true

  override def load(nbt: NBTTagCompound)
  {
    placementSide = ForgeDirection.getOrientation(nbt.getByte("side"))
    tier = nbt.getByte("tier")
  }

  override def save(nbt: NBTTagCompound)
  {
    nbt.setByte("side", placementSide.ordinal.asInstanceOf[Byte])
    nbt.setByte("tier", tier.asInstanceOf[Byte])
  }

  def getPosition: VectorWorld = new VectorWorld(world, x, y, z)

  override def toString: String = "[" + getClass.getSimpleName + "]" + x + "x " + y + "y " + z + "z " + getSlotMask + "s "
}